src/fermilib/circuits/ffft.py

```diff
@@ -0,0 +1,383 @@
+
+    # Undo the fermionic swap network to restore the original ordering.
+    for swap in all_swaps[::-1]:
+        Swap | (register[swap[0]], register[swap[1]])
+        C(Z) | (register[swap[0]], register[swap[1]])
+
+
+def swap_adjacent_fermionic_modes(fermion_operator, mode):
+    """Swap adjacent modes in a fermionic operator.
+
+    Returns: a new FermionOperator with mode and mode+1 swapped.
+
+    Args:
+        fermion_operator (projectq.FermionOperator): Original operator.
+        mode (integer): The mode to be swapped with mode + 1.
+
+    Notes:
+        Because the swap must be fermionic, the sign of the operator is
+        flipped if both creation operators mode^ and (mode+1)^ (or the
+        corresponding annihilation operators) are present.
+    """
+    new_operator = FermionOperator.zero()
+
+    for term in fermion_operator.terms:
+        new_term = list(term)
+        multiplier = 1
+        if (mode, 1) in term and (mode + 1, 1) in term:
+            multiplier *= -1
+        else:
+            if (mode, 1) in term:
+                new_term[term.index((mode, 1))] = (mode + 1, 1)
+            if (mode + 1, 1) in term:
+                new_term[term.index((mode + 1, 1))] = (mode, 1)
+
+        if (mode, 0) in term and (mode + 1, 0) in term:
+            multiplier *= -1
+        else:
+            if (mode, 0) in term:
+                new_term[term.index((mode, 0))] = (mode + 1, 0)
+            if (mode + 1, 0) in term:
+                new_term[term.index((mode + 1, 0))] = (mode, 0)
+
+        new_operator.terms[tuple(new_term)] = (multiplier *
+                                               fermion_operator.terms[term])
+
+    return new_operator
+
+
+def ffft_swap_networks(system_size, n_dimensions):
+    """Compute the two networks of swaps required for the FFFT.
+
+    Args:
+        system_size (int): The system side length (the system is assumed
+                           to be a hypercube).
+        n_dimensions (int): Number of dimensions for the system.
+
+    Notes:
+        The reordering of the operators is the opposite of the ordering
+        that would be applied to a state. This reordering is given in
+        two stages. First, the bit-reversed output of the FFFT circuit
+        (e.g. 0, 2, 1, 3 which is the four numbers 0, 1, 2, 3 after
+        reversing their bit representations) is swapped to the standard
+        order (e.g. 0, 1, 2, 3). Second, the standard order is swapped
+        to the split-halves ordering used in fourier_transform (e.g.
+        2, 3, 0, 1 which swaps the left and right halves of that range).
+
+        Both rounds must be applied before the FFFT, but after the FFFT
+        only the second round needs to be undone so as to restore to the
+        order of the FFFT circuit. All reorderings must be done in both
+        dimensions, and because the operator is fermionic all swaps must
+        also be.
+    """
+    s = '{:0' + str(int(numpy.log2(system_size))) + 'b}'
```

Comment:
Done (and the below two). The purpose of these is pretty complex and is also better commented now.